### PR TITLE
Implement CaImAn motion correction ingestion

### PR DESCRIPTION
**Pull request by kabilar**

### CODE REVIEW

**Review by ttngu207 — CHANGES_REQUESTED** [review]
Can you package this into a function, part of the `img_loader.caiman` module.
so users can invoke with something like this:

```
...
from img_loader.caimain import save_mc

mc_fp = 'data_dir/motion_correction.hdf5'
save_mc(mc,  mc_fp)
```

**Comment by kabilar** [comment]
Good idea.  Will update this pull request.  Thanks.

File: caiman_mc_save.py [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+data_dir = '/'
+
+# ---------------------------------------------------------------------------------
+import h5py
+
+filename = data_dir + '/analysis_results.hdf5'
+
+# Load motion corrected image
+mc_image = cm.load(mc.mmap_file)
+
```
Comment (ttngu207) [inline caiman_mc_save.py:17]: `Cn` can be computed here:

```
Cn = cm.local_correlations(mc_image .transpose(1,2,0))
Cn[np.isnan(Cn)] = 0
```

File: img_loaders/caiman_loader.py [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+            h5g.require_dataset("reference_image",shape=np.shape(mc.total_template_rig),data=mc.total_template_rig,dtype=mc.total_template_rig.dtype)
+
+        h5g.require_dataset("correlation_image",shape=np.shape(correlation_image),data=correlation_image,dtype=correlation_image.dtype)
+        h5g.require_dataset("average_image",shape=np.shape(average_image),data=average_image,dtype=average_image.dtype)
+        h5g.require_dataset("max_image",shape=np.shape(max_image),data=max_image,dtype=max_image.dtype)
+
+        # Close hdf5 file
+        h5f.close()
+
+    def extract_mc(self, caiman_fp):
```
Comment (ttngu207) [inline img_loaders/caiman_loader.py:143]: no need to pass in `caiman_fp` here, instead you should make use of `caiman_fp` at the `__init__` step, so users don't have to pass in `caiman_fp` twice

File: img_loaders/caiman_loader.py [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+        self.h5f = h5py.File(caiman_fp, 'r')
+
+        # ---- Declare attributes ----
+        self.dims                  = self.h5f['dims']
+        self.dview                 = self.h5f['dview']
+        self.estimates             = self.h5f['estimates']
+        self.mmap_file             = self.h5f['mmap_file']
+        self.params                = self.h5f['params']
+        self.remove_very_bad_comps = self.h5f['remove_very_bad_comps']
+        self.skip_refinement       = self.h5f['skip_refinement']
```
Comment (ttngu207) [inline img_loaders/caiman_loader.py:154]: I'd suggest to not returning these properties into `self` if there is not a need for them. The method is `extract_mc` so we can just return the `motion_correction` part.

File: img_loaders/caiman_loader.py [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+        h5g.require_dataset("correlation_image",shape=np.shape(correlation_image),data=correlation_image,dtype=correlation_image.dtype)
+        h5g.require_dataset("average_image",shape=np.shape(average_image),data=average_image,dtype=average_image.dtype)
+        h5g.require_dataset("max_image",shape=np.shape(max_image),data=max_image,dtype=max_image.dtype)
+
+        # Close hdf5 file
+        h5f.close()
+
+    def extract_mc(self, caiman_fp):
+        # ---- Open CaImAn output file ----
+        self.h5f = h5py.File(caiman_fp, 'r')
```
Comment (ttngu207) [inline img_loaders/caiman_loader.py:145]: `self.h5f` already exists (from `__init__`) so no need to recreate it here again

**Review by ttngu207 — CHANGES_REQUESTED** [review]

File: djimaging/processing.py [hunk 5]
```diff
@@ -135,27 +136,26 @@ def make(self, key):
...
-                caiman_loader = caiman.CaImAn(data_dir)
-                key = {**key, 'proc_completion_time': caiman_loader.creation_time,
-                       'proc_curation_time': caiman_loader.curation_time}
-                # Insert file(s)
-                root = pathlib.Path(PhysicalFile._get_root_data_dir())
-                output_files = data_dir.glob('*')
-                output_files = [pathlib.Path(f).relative_to(root).as_posix() for f in output_files if f.is_file()]
+                hdf5_fp = list(data_dir.glob('*.hdf5'))
+                caiman_fp = hdf5_fp[0].as_posix()
+                caiman_loader_inst = caiman_loader.CaImAn(caiman_fp)
```
Comment (ttngu207) [inline djimaging/processing.py:150]: I'd suggest passing in the `data_dir` in instantiating the caiman loader, and move the logic of searching + verifying the `hdf5` file into the loader itself

File: djimaging/processing.py [hunk 6]
```diff
@@ -135,27 +136,24 @@ def make(self, key):
...
-                output_files = [pathlib.Path(f).relative_to(root).as_posix() for f in output_files if f.is_file()]
+                caiman_loader_inst = caiman_loader.CaImAn(data_dir)
 
+                key = {**key, 'proc_completion_time': caiman_loader_inst.creation_time,
+                              'proc_curation_time': caiman_loader_inst.curation_time}
+
+                # Insert file(s)
+                # root = pathlib.Path(PhysicalFile._get_root_data_dir())
+                # output_files = [pathlib.Path(f).relative_to(root).as_posix() for f in output_files if f.is_file()]
+                output_files=[caiman_loader_inst.caiman_fp]
```
Comment (ttngu207) [inline djimaging/processing.py:156]: `caiman_loader_inst.caiman_fp` would return the absolute path of the caiman file, what we want is its relative path with respect to the user-specified `root_data_dir`, hence the extra step of doing `.relative_to(root)....` in line 155 that is commented out

File: img_loaders/caiman_loader.py [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+                          'inferred_trace': self.cnmf.estimates.C[comp_idx, :],
+                          'dff': self.cnmf.estimates.F_dff[comp_idx, :],
+                          'spikes': self.cnmf.estimates.S[comp_idx, :]})
+        return masks
+
+    def extract_mc(self):
+        self.params            = self.h5f['params']
+        self.motion_correction = self.h5f['motion_correction']
+
+        return self
```
Comment (ttngu207) [inline img_loaders/caiman_loader.py:96]: no real need to have to return `self` here, as there's already a handle on the instantiated object `CaImAn` already.
Something like:
```
cm_loader = CaImAn(path)
cm_loader.extract_mc()

cm_loader.motion_correction ...
```

**Review by ttngu207 — APPROVED** [review]

**Pull request merged**